### PR TITLE
Support underscore in numeric literals, ver 2

### LINKED ISSUE #28967: Should we support numeric literals in form of 1_000_000? (with underscore separator of groups).
Pros:
- it is convenient to type and read numbers this way;
- Ada, Perl, C++, Go and Rust already support that;
- can support thousands as well as lakhs and crores;

Cons:
- standard SQL does not have this support;
- PostgreSQL treat `SELECT 1_000_000` as number one with `_000_000` alias: http://sqlfiddle.com/#!17/0a28f/400
- C++ has `1'000'000` instead of `1_000_000`;

What do you think?

### PR DESCRIPTION
**Pull request by jh0x**
Fixes #28967

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support numeric literals with _ as separator

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by vdimir** [comment]
Thank you for your PR!

We have another opened PR with this feature, but it wasn't finished.
https://github.com/ClickHouse/ClickHouse/pull/39129

You may want to compare implementations and check out comments from that one. Maybe it will be easier to finish that code. 


Also I'm still a bit concerned about dropping support of identifiers starting with the number (`000_qwerty`). In linked PR we decided to drop them, but probably you would manage to implement it, what is the main setback for this?

**Comment by jh0x** [comment]
> You may want to compare implementations and check out comments from that one. Maybe it will be easier to finish that code.

I've done an independent implementation and kept in mind the comments. That is the reason for dropping the support for the identifiers starting with numbers.

> Also I'm still a bit concerned about dropping support of identifiers starting with the number (`000_qwerty`). In linked PR we decided to drop them, but probably you would manage to implement it, what is the main setback for this?

Actually I think there is no need for dropping them. I just pushed two commits:

- bringing back the identifiers
- restoring the changed tests to their original state

File: src/Common/StringUtils/StringUtils.h [hunk 1]
```diff
@@ -156,6 +156,20 @@ inline bool isValidIdentifier(std::string_view str)
...
 }
 
+
+inline bool isNumberSeparator(bool is_start_of_block, bool is_hex, const char* pos, const char* end)
+{
+    if (*pos != '_')
+        return false;
+    if (is_start_of_block && *pos == '_')
+        return false; // e.g. _123, 12e_3
+    if (pos + 1 < end && (pos[1] == '_' || pos[1] == 'p' || pos[1] == 'P' || pos[1] == ';' || pos[1] == '.' || (!is_hex && (pos[1] == 'e' || pos[1] == 'E'))))
```
Comment (vdimir) [inline src/Common/StringUtils/StringUtils.h:166]: Why just symbols `p`, `;`, `.`, ... ?

```
select 1_; -- Missing columns: '1_'
select 1_ ; -- returns 1
```

File: src/Parsers/ExpressionElementParsers.cpp [hunk 2]
```diff
@@ -830,21 +830,65 @@ bool ParserNumber::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
...
+
     char buf[MAX_LENGTH_OF_NUMBER + 1];
 
-    size_t size = pos->size();
-    memcpy(buf, pos->begin, size);
+    size_t buf_size = 0;
+    for (auto it = pos->begin; it != pos->end; ++it)
+    {
+        if (*it != '_')
+        buf[buf_size++] = *it;
```
Comment (vdimir) [inline src/Parsers/ExpressionElementParsers.cpp:883]:
```suggestion
    for (const auto * it = pos->begin; it != pos->end; ++it)
    {
        if (*it != '_')
            buf[buf_size++] = *it;

```

File: src/Parsers/Lexer.cpp [hunk 3]
```diff
@@ -105,44 +105,57 @@ Token Lexer::nextTokenImpl()
...
+                while (pos < end && (isNumericASCII(*pos) || isNumberSeparator(false, false, pos, end)))
                     ++pos;
             }
             else
             {
+                bool start_of_block = false;
                 /// 0x, 0b
                 bool hex = false;
-                if (pos + 2 < end && *pos == '0' && (pos[1] == 'x' || pos[1] == 'b' || pos[1] == 'X' || pos[1] == 'B'))
+                if (pos + 2 < end && *pos == '0' && (pos[1] == 'x' || pos[1] == 'b' || pos[1] == 'X' || pos[1] == 'B') && pos[2] != ';')
```
Comment (vdimir) [inline src/Parsers/Lexer.cpp:116]: Why do we check `;` ?

Comment (jh0x) [reply src/Parsers/Lexer.cpp]: Actually, I changed this logic a bit in the last of the commits. On master we had the following behaviour:
```sh
clickhouse client -q "select 0x"   # results in 0
clickhouse client -q "select 0x;"  # results in 0
clickhouse client -q "select 0x ;" # results in 0
# and same for 0b
```

That `0x` or `0b` doesn't really look like a valid number, so I changed it to the following behaviour:
```sh
clickhouse client -q "select 0x"   # results in UNKNOWN_IDENTIFIER
clickhouse client -q "select 0x;"  # results in UNKNOWN_IDENTIFIER
clickhouse client -q "select 0x ;" # results in UNKNOWN_IDENTIFIER
# and same for 0b
```

There's one more weird case, but behaviour on my branch is consistent with master:
```sql
select -_0; -- UNKNOWN_IDENTIFIER
select +_0; -- SYNTAX_ERROR
```

**Comment by vdimir** [comment]
CI failures:

> Integration tests (asan) [6/6] — fail: 25, passed: 217, flaky: 1 [Details](https://s3.amazonaws.com/clickhouse-test-reports/43925/b6015a5c020840cad71249de22ac367f750ba0a4/integration_tests__asan__[6/6].html)

Issue https://github.com/ClickHouse/ClickHouse/issues/41776

> Stateless tests (debug, s3 storage) [2/6] — fail: 7, passed: 750, skipped: 13 [Details](https://s3.amazonaws.com/clickhouse-test-reports/43925/b6015a5c020840cad71249de22ac367f750ba0a4/stateless_tests__debug__s3_storage__[2/6].html)

Timeouts with `<Warning> TCPHandler: Client has gone away`, similar report https://s3.amazonaws.com/clickhouse-test-reports/43161/9654e06ec8fa32c31aca6cf52ad3dd4f40140189/stateless_tests__debug__s3_storage__[2/6].html 

> Stateless tests (release, wide parts enabled) — fail: 1, passed: 4734, skipped: 5 [Details](https://s3.amazonaws.com/clickhouse-test-reports/43925/b6015a5c020840cad71249de22ac367f750ba0a4/stateless_tests__release__wide_parts_enabled_.html)

Issue https://github.com/ClickHouse/ClickHouse/issues/43983

> Stress test (tsan) — Backward compatibility check: Error message in clickhouse-server.log (see bc_check_error_messages.txt) [Details](https://s3.amazonaws.com/clickhouse-test-reports/43925/b6015a5c020840cad71249de22ac367f750ba0a4/stress_test__tsan_.html)

Issue https://github.com/ClickHouse/ClickHouse/issues/41369

> Stress test (ubsan) — Fatal message in clickhouse-server.log (see fatal_messages.txt) [Details](https://s3.amazonaws.com/clickhouse-test-reports/43925/b6015a5c020840cad71249de22ac367f750ba0a4/stress_test__ubsan_.html)

Already fixed in https://github.com/ClickHouse/ClickHouse/pull/44045

**Review by vdimir — APPROVED** [review]
@jh0x @unbyte Thank you for your effort!

**Comment by crakjie** [comment]
Nice feature :+1:  .
Have you considerate space ? `10000`  =>  `10 000`

**Pull request merged**